### PR TITLE
Do not store api key in retry file

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?

This PR avoids storing the API key in `.retry` files.

### Additional Notes

If `agentchecks` transactions can use `TransactionPriorityHigh`, then `SubmitAgentchecksMetadata` and `SendAgentchecksMetadata` can be removed.

### Describe your test plan
- Set `forwarder_retry_queue_payloads_max_size` to 50
- Set `forwarder_storage_max_size_in_bytes` to `50000000`
- Run the Agent so that there are more than 1 hour of `retry` files (to make sure all metadata requests are sent)
- Check the API key is not stored in any `.retry` files (Also check API Key is not stored in the compressed payload). I can provide a tool to dump the `retry_file` to json.
- Check the domain is not stored in any `.retry` files.
- Check `HostMetadata ` and `AgentChecks` are not stored on disk. 
- Check a `.retry` file can be reloaded on Windows.
- Check the behavior when decreasing the number of API keys for the same domain.
- Check `.retry` file cannot be edited by a regular user (Windows + Unix).

### CODE REVIEW

File: pkg/forwarder/forwarder.go [hunk 1]
```diff
@@ -551,8 +568,15 @@ func (f *DefaultForwarder) SubmitV1CheckRuns(payload Payloads, extra http.Header
...
-	transactions := f.createPriorityHTTPTransactions(v1IntakeEndpoint, payload, true, extra, priority)
+func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) error {
+	return f.submitV1Intake(payload, extra, f.createHTTPTransactions)
+}
+
+func (f *DefaultForwarder) submitV1Intake(
+	payload Payloads,
+	extra http.Header,
+	createHTTPTransactions func(endpoint endpoint, payload Payloads, apiKeyInQueryString bool, extra http.Header) []*HTTPTransaction) error {
+	transactions := createHTTPTransactions(v1IntakeEndpoint, payload, true, extra)
```
Comment (truthbk) [inline pkg/forwarder/forwarder.go:579]: So, maybe I'm missing something here.... Originally `SubmitMetadata()` and `SubmitHostMetadata()` were submitting to `metadataEndpoint`, but with this change we now use `submitV1Intake(...)` which will  create the transaction with destination endpoint  `v1IntakeEndpoint`. Now, I'm not sure what the current spec is for that endpoint but is this change intended?

File: pkg/forwarder/forwarder.go [hunk 2]
```diff
@@ -139,14 +139,15 @@ type Forwarder interface {
...
-	SubmitV1Intake(payload Payloads, extra http.Header, priority TransactionPriority) error
+	SubmitV1Intake(payload Payloads, extra http.Header) error
 	SubmitV1CheckRuns(payload Payloads, extra http.Header) error
 	SubmitSeries(payload Payloads, extra http.Header) error
 	SubmitEvents(payload Payloads, extra http.Header) error
 	SubmitServiceChecks(payload Payloads, extra http.Header) error
 	SubmitSketchSeries(payload Payloads, extra http.Header) error
 	SubmitHostMetadata(payload Payloads, extra http.Header) error
-	SubmitMetadata(payload Payloads, extra http.Header, priority TransactionPriority) error
+	SubmitAgentchecksMetadata(payload Payloads, extra http.Header) error
```
Comment (truthbk) [inline pkg/forwarder/forwarder.go:149]: nit: I think we might want `SubmitAgentChecksMetadata()` instead for consistent CamelCase.

File: pkg/forwarder/sync_forwarder.go [hunk 3]
```diff
@@ -97,14 +97,17 @@ func (f *SyncForwarder) SubmitSketchSeries(payload Payloads, extra http.Header)
 
 // SubmitHostMetadata will send a host_metadata tag type payload to Datadog backend.
 func (f *SyncForwarder) SubmitHostMetadata(payload Payloads, extra http.Header) error {
-	transactions := f.defaultForwarder.createHTTPTransactions(hostMetadataEndpoint, payload, false, extra)
-	return f.sendHTTPTransactions(transactions)
+	return f.SubmitV1Intake(payload, extra)
```
Comment (truthbk) [inline pkg/forwarder/sync_forwarder.go:100]: Same as mentioned in the forwarder, this changes the destination endpoint. Intended?

File: pkg/forwarder/transaction.go [hunk 4]
```diff
@@ -381,5 +385,9 @@ func (t *HTTPTransaction) internalProcess(ctx context.Context, client *http.Clie
 
 // SerializeTo serializes the transaction using TransactionsSerializer
 func (t *HTTPTransaction) SerializeTo(serializer *TransactionsSerializer) error {
-	return serializer.Add(t)
+	if t.storableOnDisk {
+		return serializer.Add(t)
+	}
+	log.Debug("The transaction is not stored on disk because `storableOnDisk` is false.")
```
Comment (truthbk) [inline pkg/forwarder/transaction.go:391]: Is this too verbose? Maybe more a trace-level statement? (I'm not sure just thinking out loud). Also, this doesn't tell us much about the transaction that was not stored to disk... Anyway, please let me known your thoughts.

File: pkg/forwarder/transactions_serializer.go [hunk 5]
```diff
@@ -47,9 +71,9 @@ func (s *TransactionsSerializer) Add(transaction *HTTPTransaction) error {
 
 	endpoint := transaction.Endpoint
 	transactionProto := HttpTransactionProto{
-		Domain:     transaction.Domain,
-		Endpoint:   &EndpointProto{Route: endpoint.route, Name: endpoint.name},
-		Headers:    toHeaderProto(transaction.Headers),
+		Domain:     "", // Domain is always the same
```
Comment (truthbk) [inline pkg/forwarder/transactions_serializer.go:74]: Even though it's the same domain, why not include it? To save space? Do we compress when we write to disk? If we compress the payloads, then including the `Domain` should have negligible impact.

**Review by truthbk — COMMENTED** [review]
Very nice work, just a few questions and couple nits. Let me know if I missed anything, which I definitely may have 😅

File: pkg/forwarder/forwarder.go (hunk 1 above)
Comment (ogaca-dd) [reply pkg/forwarder/forwarder.go]: I think these updates come from https://github.com/DataDog/datadog-agent/pull/7077.

> Note: SubmitMetadata and SubmitHostMetadata were unused.
Before, Serializer.SendMetadata and Serializer.SendHostMetadata called Forwarder.SubmitV1Intake.
Now these 2 functions call respectively Forwarder.SubmitMetadata and Forwarder.SubmitHostMetadata which perform the same as Forwarder.SubmitV1Intake (Forwarder.SubmitHostMetadata use TransactionPriorityHigh)


Note: My target branch was originally `olivierg/simplify-forwarder-interface`.
Let me know if anything is unclear.

File: pkg/forwarder/transactions_serializer.go (hunk 5 above)
Comment (ogaca-dd) [reply pkg/forwarder/transactions_serializer.go:74]: It is mostly for security reasons. If a user can update the domain from the serialized transactions, they can replace the domain by `127.0.0.1:1234` and then the Agent will send the HTTP transactions to `127.0.0.1:1234` on an endpoint containing the API KEY.

File: pkg/util/common/slice.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package common
+
+// StringSliceMap returns a new slice `s` where s[i] = fct(values[i])
+func StringSliceMap(values []string, fct func(string) string) []string {
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/util/common/slice.go:9]: Updated.

File: pkg/forwarder/transactions_serializer.go [hunk 7]
```diff
@@ -69,42 +93,77 @@ func (s *TransactionsSerializer) GetBytesAndReset() ([]byte, error) {
...
+		if err == nil {
+			route, err = s.restoreAPIKeys(e.Route)
+			if err == nil {
+				proto, err = s.fromHeaderProto(transaction.Headers)
+			}
+		}
+
 		if err != nil {
-			return nil, err
+			log.Error(err)
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply pkg/forwarder/transactions_serializer.go:119]: Updated.

File: pkg/forwarder/transaction.go (hunk 4 above)
Comment (ogaca-dd) [reply pkg/forwarder/transaction.go:393]: I have changed from `Debug` to `Trace` (it can be useful for the QA of this development).

**Review by truthbk — APPROVED** [review]
Excellent stuff! 🙇

File: pkg/forwarder/transactions_serializer.go (hunk 5 above)
Comment (olivielpeau) [reply pkg/forwarder/transactions_serializer.go:74]: I think this should be explained in a comment inline and tested (I didn't see a test for this but may have missed something), otherwise this crucial context will be lost in the future
